easybuild/easyconfigs/j/jax/jax-0.4.29-gfbf-2023b-CUDA-12.4.0.eb

```diff
@@ -0,0 +1,133 @@
+    ('zlib', '1.2.13'),
+    ('Python', '3.11.5'),
+    ('SciPy-bundle', '2023.11'),
+    ('flatbuffers-python', '23.5.26'),
+    ('ml_dtypes', '0.4.0'),
+]
+
+# downloading xla and other tarballs to avoid that Bazel downloads it during the build
+# note: this *must* be the exact same commit as used in third_party/{xla,"other"}/workspace.bzl
+local_xla_commit = '32ba408c0e2b7ef5f5821c0781601ba17d467076'
+#local_tfrt_commit = '0aeefb1660d7e37964b2bb71b1f518096bda9a25'
+
+local_extract_cmd = 'mkdir -p %(builddir)s/archives && cp %s %(builddir)s/archives'
+local_repo_opt = '--bazel_options="--distdir=%(builddir)s/archives" '
+local_repo_opt += '--bazel_options="--action_env=TF_SYSTEM_LIBS=pybind11" '
+local_repo_opt += '--bazel_options="--action_env=CPATH=$EBROOTPYBIND11/include" '
+
+
+# Some tests require an isolated run:
+local_isolated_tests = [
+    'tests/host_callback_test.py::HostCallbackTapTest::test_tap_scan_custom_jvp',
+    'tests/host_callback_test.py::HostCallbackTapTest::test_tap_transforms_doc',
+    'tests/lax_scipy_special_functions_test.py::LaxScipySpcialFunctionsTest' +
+    '::testScipySpecialFun_gammainc_s_2x1x4_float32_float32',
+]
+# deliberately not testing in parallel, as that results in (additional) failing tests;
+# use XLA_PYTHON_CLIENT_ALLOCATOR=platform to allocate and deallocate GPU memory during testing,
+# see https://github.com/google/jax/issues/7323 and
+# https://github.com/google/jax/blob/main/docs/gpu_memory_allocation.rst;
+# use CUDA_VISIBLE_DEVICES=0 to avoid failing tests on systems with multiple GPUs;
+# use NVIDIA_TF32_OVERRIDE=0 to avoid loosing numerical precision by disabling TF32 Tensor Cores;
+local_test_exports = [
+    "NVIDIA_TF32_OVERRIDE=0",
+    "CUDA_VISIBLE_DEVICES=0",
+    "XLA_PYTHON_CLIENT_ALLOCATOR=platform",
+    "JAX_ENABLE_X64=true",
+]
+local_test = ''.join(['export %s;' % x for x in local_test_exports])
+# run all tests at once except for local_isolated_tests:
+local_test += "pytest -vv tests %s && " % ' '.join(['--deselect %s' % x for x in local_isolated_tests])
+# run remaining local_isolated_tests separately: 
+local_test += ' && '.join(['pytest -vv %s' % x for x in local_isolated_tests])
+
+use_pip = True
+
+default_easyblock = 'PythonPackage'
+default_component_specs = {
+    'sources': [SOURCE_TAR_GZ],
+    'source_urls': [PYPI_SOURCE],
+    'start_dir': '%(name)s-%(version)s',
+    'use_pip': True,
+    'sanity_pip_check': True,
+    'download_dep_fail': True,
+}
+
+components = [
+    ('absl-py', '2.1.0', {
+        'options': {'modulename': 'absl'},
+    }),
+    ('jaxlib', version, {
+        'sources': [
+            '%(name)s-v%(version)s.tar.gz',
+            {
+                'download_filename': '%s.tar.gz' % local_xla_commit,
+                'filename': 'xla-%s.tar.gz' % local_xla_commit,
+                'extract_cmd': local_extract_cmd,
+            },
+#            {
+#                'download_filename': '%s.tar.gz' % local_tfrt_commit,
```

Comment:
The build log (without this) doesn't show any traces of downloading tensorflow, in fact there is no "tensorflow" in it at all.
Compared to jax-0.3.25-foss-2022a.eb with the explicit download of TF removed, which does show it downloading TF and lots of references to the downloaded tensorflow during building.

So I'd guess that it isn't used any longer.

Reply:
Note that this isn't tensorflow but a library they extracted from it to reuse in XLA called [tfrt](https://github.com/tensorflow/runtime/) in Bazel. Have you checked the bazel cache for "tf_runtime" or "tfrt" files/folders?

To me it looks like it is still used comparing to the previous EC:
- Jax 0.4.25 uses XLA [4ccfe33c71665ddcbca5b127fefe8baa3ed632d4](https://github.com/google/jax/blob/jaxlib-v0.4.25/third_party/xla/workspace.bzl#L23)
- That XLA uses TFRT [0aeefb1660d7e37964b2bb71b1f518096bda9a25](https://github.com/openxla/xla/blob/4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/third_party/tsl/third_party/tf_runtime/workspace.bzl#L9)

And now similar:
- jax 0.4.29 uses XLA [32ba408c0e2b7ef5f5821c0781601ba17d467076](https://github.com/google/jax/blob/jaxlib-v0.4.29/third_party/xla/workspace.bzl#L23)
- That XLA uses TFRT [4662f7552dc2e420ee20361c077b7aeb334a1087](https://github.com/openxla/xla/blob/32ba408c0e2b7ef5f5821c0781601ba17d467076/third_party/tsl/third_party/tf_runtime/workspace.bzl#L9)

Given that Bazel downloads much more it might not be worth adding one more or less. But still: The more we have the better. Maybe also add the URLs I linked above as comments to where the commit hashes come from

Otherwise I think it's better to just remove the commented code part here.

Reply:
I made two builds, one without explicit download and one with explicit download.
I can't find any trace whatsoever of it being used (or downloaded when not doing it explicitly) in either build.

It might be that XLA needs a flag to enable it?

And by the looks of it, in both cases it downloads xla by itself, at least I can find it in the bazel cache in both cases.

Reply:
> It might be that XLA needs a flag to enable it?

Then out would be strange that this changed after the last EC. At least the linked code looks similar enough

>  I can't find any trace whatsoever of it being used (or downloaded when not doing it explicitly) in either build.

Hm... Currently can't check myself. I'd expect it somewhere in "bazel-out", could do after my vacation, so in 1.5 weeks

> And by the looks of it, in both cases it downloads xla by itself, at least I can find it in the bazel cache in both cases.

That is expected: Bazel checks the `distdir` and downloads things if not found. In either case it puts it into the cache. I guess to not do the check again and possibly modify it. IIRC it might apply patches but won't ever modify the original source (i.e. ours in distdir)

Reply:
I went through all sha256 files in the bazel cache and non of them contains tfrt in any of my three builds.

And it does look like it's not using it in this version of xla, there is no mention of tf_runtime or tfrt_dependencies in third_party/tsl/workspace*bzl. They are there in an older xla version (i.e. the one used in jax 0.4.25)

Reply:
Comparing grep output:

```
grep -r tfrt_dependencies xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/third_party/tsl/workspace2.bzl:load("@tf_runtime//:dependencies.bzl", "tfrt_dependencies")
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/third_party/tsl/workspace2.bzl:    tfrt_dependencies()
```
vs
```
grep -r tfrt_dependencies xla-32ba408c0e2b7ef5f5821c0781601ba17d467076/
```
and

```
grep -r tf_runtime xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/third_party/tsl/opensource_only.files:third_party/tf_runtime/BUILD:
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/third_party/tsl/third_party/tf_runtime/workspace.bzl:        name = "tf_runtime",
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/third_party/tsl/workspace2.bzl:load("@tf_runtime//:dependencies.bzl", "tfrt_dependencies")
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/third_party/tsl/workspace3.bzl:load("//third_party/tf_runtime:workspace.bzl", tf_runtime = "repo")
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/third_party/tsl/workspace3.bzl:    tf_runtime()
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/xla/mlir/math/BUILD:        # "//third_party/tf_runtime/...",
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/xla/mlir/memref/BUILD:        # "//third_party/tf_runtime/...",
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/xla/mlir/runtime/BUILD:        # "//third_party/tf_runtime/...",
xla-4ccfe33c71665ddcbca5b127fefe8baa3ed632d4/xla/mlir/runtime/BUILD:        # "//third_party/tf_runtime_google/...",
```
vs
```
grep -r tf_runtime xla-32ba408c0e2b7ef5f5821c0781601ba17d467076/
xla-32ba408c0e2b7ef5f5821c0781601ba17d467076/third_party/tsl/opensource_only.files:third_party/tf_runtime/BUILD:
xla-32ba408c0e2b7ef5f5821c0781601ba17d467076/third_party/tsl/third_party/tf_runtime/workspace.bzl:        name = "tf_runtime",
```

So defined as repo but never accessed as far as I can see.

Reply:
Yep, I'd say too that given that `tf_runtime()` is missing it doesn't load the repo. So :+1: from me